libxrdp/xrdp_bitmap32_compress.c

```diff
@@ -459,35 +461,44 @@ xrdp_bitmap32_compress(char *in_data, int width, int height,
             fdelta(sr_data, r_data, cx, cy);
             fdelta(sg_data, g_data, cx, cy);
             fdelta(sb_data, b_data, cx, cy);
-            out_uint8(s, header);
-            r_bytes = fpack(r_data, cx, cy, s);
-            g_bytes = fpack(g_data, cx, cy, s);
-            b_bytes = fpack(b_data, cx, cy, s);
-            total_bytes = r_bytes + g_bytes + b_bytes;
-            if (1 + total_bytes > byte_limit)
+            while (cy > 0)
             {
-                /* failed */
-                LLOGLN(0, ("xrdp_bitmap32_compress: too big, rgb "
-                       "bytes %d %d %d total_bytes %d cx %d cy %d "
-                       "byte_limit %d", r_bytes, g_bytes, b_bytes,
-                       total_bytes, cx, cy, byte_limit));
-                return 0;
-            }
-            max_bytes = cx * cy * 3;
-            if (total_bytes > max_bytes)
-            {
-                /* raw is better */
-                LLOGLN(10, ("xrdp_bitmap32_compress: too big, rgb "
-                       "bytes %d %d %d total_bytes %d cx %d cy %d "
-                       "max_bytes %d", r_bytes, g_bytes, b_bytes,
-                       total_bytes, cx, cy, max_bytes));
-                init_stream(s, 0);
-                foutraw3(s, cx * cy, FLAGS_NOALPHA, sr_data, sg_data, sb_data);
+                s->p = hold_p;
+                out_uint8(s, header);
+                r_bytes = fpack(r_data, cx, cy, s);
+                g_bytes = fpack(g_data, cx, cy, s);
+                b_bytes = fpack(b_data, cx, cy, s);
+                max_bytes = cx * cy * 3;
+                total_bytes = r_bytes + g_bytes + b_bytes;
+                if (total_bytes > max_bytes)
+                {
+                    if (2 + max_bytes <= byte_limit)
+                    {
+                        s->p = hold_p;
+                        foutraw3(s, cx * cy, FLAGS_NOALPHA, sr_data, sg_data, sb_data);
```

Comment:
I would add checks to the functions like `foutraw3` instead of doing those checks in the callers. The function could also print debug information if the data is being dropped. If any caller doesn't been the check, it's always possible to define `foutraw3_unchecked`.